pkg/proxy/backend/authenticator.go

```diff
@@ -101,8 +105,9 @@ func (auth *Authenticator) verifyBackendCaps(logger *zap.Logger, backendCapabili
 	return nil
 }
 
-func (auth *Authenticator) handshakeFirstTime(logger *zap.Logger, clientIO *pnet.PacketIO, handshakeHandler HandshakeHandler,
-	getBackend backendIOGetter, frontendTLSConfig, backendTLSConfig *tls.Config) error {
+func (mgr *BackendConnManager) handshakeFirstTime(logger *zap.Logger, clientIO, backendIO *pnet.PacketIO, handshakeHandler HandshakeHandler, frontendTLSConfig, backendTLSConfig *tls.Config) error {
```

Comment:
I don't like this change.
The code is all about authentication but you moved it to `BackendConnManager`.
Besides, it's also not a good idea to introduce the concept `router` in the authentication.

Reply:
Considering that they are actually coded closely despite the concepts are separated and clear, I prefer integrating them together.

Here are some arguments:

1. `authenticator` actually holds the customized context of `handshakeHandler` anyway.
2. It contains proxy protocol. And it can only be done in the authenticator.
3. `handshakeHandler` is passed to customize `HandshakeResp`. To archive the full logic of serverless gateway, maybe there will be more customization.
4. Actually `BackendConnMgr` depends on internal fields of authenticator, e.g. capability, which should be connection states instead of authenticator-only information, IMO.

But I agree that maybe this function is too long. One solution would be moving the `getBackendIO` into a separate function, then it still looks like the old code. The only difference would be the function signature.

Reply:
I would say that the problem is in MYSQL protocol. The handshake process itself is not all about authentication, but also load balance, server capability, session tracking, TLS security, compression, etc... Those are not part of the authenticator. Instead, authentication is part of the handshake.

Reply:
You can rename `Authenticator` to `Handshaker` or whatever that is related to handshake, but it processes all the things about handshaking. `BackendConnManager` splits some logic to `Authenticator` and `CmdProcessor` to make itself simpler.
If you weaken the boundary of these logics, it again makes them coupled and even into one struct finally.

Reply:
If it's all about readability, you can pass a `HandshakeHandlerWrapper` to `Authenticator`, which adds a `GetBackendIO()` function to `HandshakeHandler`.

Reply:
I really don't mind it is one struct or two. There are lots of mega structs in various projects. I can accept something like:

```
type CommandProcessor struct {
   ...
}

type Conn struct {
  // common fields
  ...
  // authentication only fields
  ...
  // command processor fields
  cmdprc *CommandProcessor
}
```

I cares more about if the definition of functions are split accordingly. That is no coupling for me.

Reply:
What about now? Now `handshakeFirstTime` is still a method of `BackendConnMgr` located at `authenticator.go` and invoke `mgr.getBackendIO()` to do the routing. The only difference is the function signature.